File: test/inttypes.c
/* Behavioral contract for the numeric kernels behind int/s64 and int/u64, run
 * against whichever implementation the build selected (`-Dint-types-core=c` or
 * the Zig default). */

#include <assert.h>
#include <math.h>
#include <stdint.h>
#include <string.h>
#include <janet.h>

static JanetFunction *compare_fn;

/* Compare through a compiled function rather than a literal expression, so a
 * double argument never becomes a compile-time constant. janetc_loadconst
 * casts a NaN constant to int32_t without excluding NaN first -- see
 * FOUND.md -- and these vectors must not depend on that unresolved defect. */
static double compare_values(Janet a, Janet b) {
    Janet argv[2];
    Janet out;
    argv[0] = a;
    argv[1] = b;
    assert(janet_pcall(compare_fn, 2, argv, &out, NULL) == JANET_SIGNAL_OK);
    assert(janet_checktype(out, JANET_NUMBER));
    return janet_unwrap_number(out);
}

static double compare_s64_double(int64_t x, double y) {
    return compare_values(janet_wrap_s64(x), janet_wrap_number(y));
}

static double compare_u64_double(uint64_t x, double y) {
    return compare_values(janet_wrap_u64(x), janet_wrap_number(y));
}

static void test_hash(void) {
    int64_t a = 0;
    int64_t b = 1;
    int64_t c = INT64_MIN;
    uint64_t u = 1;

    /* The hash folds the two halves together, so it is stable and independent
     * of which of the two types holds the bits. */
    assert(janet_s64_type.hash(&a, sizeof(a)) == 0);
    assert(janet_s64_type.hash(&b, sizeof(b)) == janet_u64_type.hash(&u, sizeof(u)));
    assert(janet_s64_type.hash(&c, sizeof(c)) == INT32_MIN);
    assert(janet_s64_type.hash(&b, sizeof(b)) != janet_s64_type.hash(&a, sizeof(a)));

    /* Values differing only in the high word still separate. */
    {
        int64_t high = ((int64_t) 1) << 32;
        assert(janet_s64_type.hash(&high, sizeof(high)) == 1);
    }
}

static void test_abstract_compare(void) {
    int64_t s_small = -5;
    int64_t s_big = 5;
    int64_t s_min = INT64_MIN;
    int64_t s_max = INT64_MAX;
    uint64_t u_small = 5;
    uint64_t u_big = UINT64_MAX;

    assert(janet_s64_type.compare(&s_small, &s_big) == -1);
    assert(janet_s64_type.compare(&s_big, &s_small) == 1);
    assert(janet_s64_type.compare(&s_big, &s_big) == 0);
    assert(janet_s64_type.compare(&s_min, &s_max) == -1);
    assert(janet_s64_type.compare(&s_max, &s_min) == 1);

    /* The unsigned comparison must not borrow the signed ordering. */
    assert(janet_u64_type.compare(&u_small, &u_big) == -1);
    assert(janet_u64_type.compare(&u_big, &u_small) == 1);
    assert(janet_u64_type.compare(&u_big, &u_big) == 0);
    {
        uint64_t high_bit = ((uint64_t) 1) << 63;
        uint64_t one = 1;
        assert(janet_u64_type.compare(&high_bit, &one) == 1);
    }
}

static void test_compare_s64_double(void) {
    /* Inside the double's contiguous integer range the comparison is exact. */
    assert(compare_s64_double(0, 0.0) == 0);
    assert(compare_s64_double(5, 5.0) == 0);
    assert(compare_s64_double(5, 5.5) == -1);
    assert(compare_s64_double(6, 5.5) == 1);
    assert(compare_s64_double(-5, -5.0) == 0);
    assert(compare_s64_double(-6, -5.5) == -1);
    assert(compare_s64_double(-5, -5.5) == 1);

    /* NaN compares equal to everything, which is how the C original reports
     * "no ordering" through an int return. */
    assert(compare_s64_double(0, NAN) == 0);
    assert(compare_s64_double(INT64_MAX, NAN) == 0);

    /* Infinities sit outside every integer. */
    assert(compare_s64_double(INT64_MAX, INFINITY) == -1);
    assert(compare_s64_double(INT64_MIN, -INFINITY) == 1);

    /* Beyond 2^53 the integer can no longer be widened without rounding, so
     * the double is narrowed instead. */
    assert(compare_s64_double(INT64_MAX, 1e300) == -1);
    assert(compare_s64_double(INT64_MIN, -1e300) == 1);
    assert(compare_s64_double(INT64_MAX, 9.3e18) == -1);
    assert(compare_s64_double(INT64_MIN, -9.3e18) == 1);

    /* 2^53 itself is the edge of the exact range. */
    assert(compare_s64_double(9007199254740992LL, 9007199254740992.0) == 0);
    assert(compare_s64_double(9007199254740993LL, 9007199254740992.0) == 1);
    assert(compare_s64_double(-9007199254740993LL, -9007199254740992.0) == -1);
}

static void test_compare_u64_double(void) {
    assert(compare_u64_double(0, 0.0) == 0);
    assert(compare_u64_double(5, 5.0) == 0);
    assert(compare_u64_double(5, 5.5) == -1);
    assert(compare_u64_double(6, 5.5) == 1);

    /* Every unsigned value is above every negative double, including zero
     * against a small negative. */
    assert(compare_u64_double(0, -0.5) == 1);
    assert(compare_u64_double(0, -1e300) == 1);
    assert(compare_u64_double(UINT64_MAX, -1.0) == 1);

    assert(compare_u64_double(0, NAN) == 0);
    assert(compare_u64_double(UINT64_MAX, NAN) == 0);
    assert(compare_u64_double(UINT64_MAX, INFINITY) == -1);

    assert(compare_u64_double(UINT64_MAX, 1e300) == -1);
    assert(compare_u64_double(9007199254740992ULL, 9007199254740992.0) == 0);
    assert(compare_u64_double(9007199254740993ULL, 9007199254740992.0) == 1);
}

static void test_compare_across_types(void) {
    Janet s_neg = janet_wrap_s64(-1);
    Janet s_zero = janet_wrap_s64(0);
    Janet s_max = janet_wrap_s64(INT64_MAX);
    Janet u_zero = janet_wrap_u64(0);
    Janet u_small = janet_wrap_u64(1);
    Janet u_huge = janet_wrap_u64(((uint64_t) INT64_MAX) + 1);
    Janet u_max = janet_wrap_u64(UINT64_MAX);

    /* A negative signed value is below every unsigned value. */
    assert(compare_values(s_neg, u_zero) == -1);
    assert(compare_values(s_neg, u_max) == -1);
    assert(compare_values(u_zero, s_neg) == 1);
    assert(compare_values(u_max, s_neg) == 1);

    /* An unsigned value above INT64_MAX is above every signed value. */
    assert(compare_values(s_max, u_huge) == -1);
    assert(compare_values(u_huge, s_max) == 1);
    assert(compare_values(s_max, u_max) == -1);

    /* Inside the overlap the ordering is ordinary. */
    assert(compare_values(s_zero, u_zero) == 0);
    assert(compare_values(s_zero, u_small) == -1);
    assert(compare_values(u_small, s_zero) == 1);
    assert(compare_values(s_max, janet_wrap_u64((uint64_t) INT64_MAX)) == 0);
}

static void test_tostring(void) {
    JanetBuffer *buffer = janet_buffer(0);
    int64_t s;
    uint64_t u;

    s = 0;
    janet_s64_type.tostring(&s, buffer);
    assert(buffer->count == 1 && !memcmp(buffer->data, "0", 1));

    buffer->count = 0;
    s = INT64_MIN;
    janet_s64_type.tostring(&s, buffer);
    assert(buffer->count == 20);
    assert(!memcmp(buffer->data, "-9223372036854775808", 20));

    buffer->count = 0;
    s = INT64_MAX;
    janet_s64_type.tostring(&s, buffer);
    assert(buffer->count == 19);
    assert(!memcmp(buffer->data, "9223372036854775807", 19));

    /* The unsigned formatter must not print the high bit as a sign. */
    buffer->count = 0;
    u = UINT64_MAX;
    janet_u64_type.tostring(&u, buffer);
    assert(buffer->count == 20);
    assert(!memcmp(buffer->data, "18446744073709551615", 20));

    /* Formatting appends rather than replacing. */
    buffer->count = 0;
    janet_buffer_push_cstring(buffer, "n=");
    u = 42;
    janet_u64_type.tostring(&u, buffer);
    assert(buffer->count == 4 && !memcmp(buffer->data, "n=42", 4));
}

/* Floored division and modulo, exercised through the `div` and `mod` methods.
 *
 * INT64_MIN with a divisor of -1 is deliberately absent. The `/` and `%`
 * methods reject it with a Janet error, but `div` and `mod` do not guard it and
 * reach an undefined division in C -- see FOUND.md. Pinning either outcome
 * would assert behavior that is not yet decided. */
static void test_divf_mod(void) {
    JanetTable *env = janet_core_env(NULL);
    Janet result;

    struct {
        const char *source;
        const char *expected;
    } cases[] = {
        /* Floored division rounds toward negative infinity, unlike `/`. */
        {"(div (int/s64 7) (int/s64 2))", "3"},
        {"(div (int/s64 -7) (int/s64 2))", "-4"},
        {"(div (int/s64 7) (int/s64 -2))", "-4"},
        {"(div (int/s64 -7) (int/s64 -2))", "3"},
        {"(div (int/s64 8) (int/s64 2))", "4"},
        {"(div (int/s64 -8) (int/s64 2))", "-4"},
        {"(div (int/s64 0) (int/s64 -3))", "0"},
        /* rdiv swaps the operands. */
        {"(:rdiv (int/s64 2) (int/s64 -7))", "-4"},
        /* Floored modulo takes the sign of the divisor, unlike `%`. */
        {"(mod (int/s64 7) (int/s64 2))", "1"},
        {"(mod (int/s64 -7) (int/s64 2))", "1"},
        {"(mod (int/s64 7) (int/s64 -2))", "-1"},
        {"(mod (int/s64 -7) (int/s64 -2))", "-1"},
        {"(mod (int/s64 -8) (int/s64 2))", "0"},
        {"(% (int/s64 -7) (int/s64 2))", "-1"},
        /* A zero divisor returns the dividend rather than raising. */
        {"(mod (int/s64 7) (int/s64 0))", "7"},
        {"(mod (int/s64 -7) (int/s64 0))", "-7"},
        {"(:rmod (int/s64 2) (int/s64 -7))", "1"},
        /* The extremes still divide when the result is representable. */
        {"(div (int/s64 \"-9223372036854775808\") (int/s64 2))", "-4611686018427387904"},
        {"(mod (int/s64 \"-9223372036854775808\") (int/s64 3))", "1"},
    };

    for (size_t i = 0; i < sizeof(cases) / sizeof(cases[0]); i++) {
        assert(janet_dostring(env, cases[i].source, "test", &result) == 0);
        assert(janet_is_int(result) == JANET_INT_S64);
        {
            JanetBuffer *buffer = janet_buffer(0);
            janet_s64_type.tostring(janet_unwrap_abstract(result), buffer);
            assert(buffer->count == (int32_t) strlen(cases[i].expected));
            assert(!memcmp(buffer->data, cases[i].expected, buffer->count));
        }
    }

    /* Dividing by zero raises, while the modulo above does not. Run it under a
     * protected call so the expected failure stays off stderr. */
    assert(janet_dostring(env, "(fn [] (div (int/s64 1) (int/s64 0)))", "test", &result) == 0);
    {
        Janet out;
        assert(janet_pcall(janet_unwrap_function(result), 0, NULL, &out, NULL) == JANET_SIGNAL_ERROR);
    }
}

int main(void) {
    JanetTable *env;
    Janet fn;

    janet_init();
    env = janet_core_env(NULL);
    assert(janet_dostring(env, "(fn [a b] (compare a b))", "test", &fn) == 0);
    compare_fn = janet_unwrap_function(fn);
    janet_gcroot(fn);

    test_hash();
    test_abstract_compare();
    test_compare_s64_double();
    test_compare_u64_double();
    test_compare_across_types();
    test_tostring();
    test_divf_mod();

    janet_gcunroot(fn);
    janet_deinit();
    return 0;
}

File: test/inttypes.zig
//! Behavioral contract for the numeric kernels behind `int/s64` and
//! `int/u64`: the hash, the two abstract comparisons, the three mixed-type
//! orderings, the formatters, and floored division.
//!
//! ## The abstract types are reached as themselves
//!
//! The two `AbstractType` values have a raising `tostring`, and this file
//! calls it with `try` over the runtime's own type rather than through
//! anything that would flatten the raise.
//!
//! ## Why comparison goes through a compiled Janet function
//!
//! A double argument written as a Janet literal becomes a compile-time
//! constant, so a contract that wrote its vectors as literals would be
//! asserting about the constant folder as much as about `compare`. Calling a
//! compiled `(fn [a b] (compare a b))` with values built in Zig keeps the
//! folder out of it, so these vectors do not depend on it.
//!
//! ## NaN compares equal to everything
//!
//! That is not a bug being pinned; it is how a comparison restricted to `-1`,
//! `0` and `1` reports "no ordering". It is asserted so that a change to it
//! would be deliberate.

// ==========================================================================
// Standard library imports
// ==========================================================================

const std = @import("std");

// ==========================================================================
// Project imports
// ==========================================================================

const abi = @import("abi");
const buffers = @import("subsystems").value.buffers;
const constants = @import("constants");
const core_env = @import("subsystems").env;
const expect = @import("expect.zig").expect;
const functions = @import("subsystems").value.functions;
const gc_alloc = @import("subsystems").gc_alloc;
const harness = @import("harness.zig");
const inttypes = @import("subsystems").inttypes;
const repr = @import("repr");
const strings = @import("subsystems").value.strings;
const subsystems = @import("subsystems");
const tables = @import("subsystems").value.tables;
const vm_entry = @import("subsystems").vm_entry;
const vm_lifecycle = @import("subsystems").lifecycle;
const wrap = @import("subsystems").value.wrap;

// ==========================================================================
// Constants
// ==========================================================================

var compare_fn: *functions.Function = undefined;
var environment: *tables.Table = undefined;

/// The runtime's two abstract types, by import.
///
/// Pointers rather than aliases: an alias of a `const` is a *copy*, so an
/// address taken through one would be an address the runtime never handed out,
/// and an abstract built with it would be refused by its own type.
const s64_type = &inttypes.s64Type;
const u64_type = &inttypes.u64Type;

// ==========================================================================
// Aliased types
// ==========================================================================

const AbstractType = subsystems.abstract_type.AbstractType;

// ==========================================================================
// Cases
// ==========================================================================

fn compareValues(a: repr.Value, b: repr.Value) f64 {
    var argv = [2]repr.Value{ a, b };
    const resumed = vm_entry.pcall(compare_fn, &argv, null);
    expect(resumed.signal == abi.Signal.ok);
    expect(harness.isType(resumed.value, repr.Tag.number));
    return wrap.toNumber(resumed.value);
}

fn compareS64Double(x: i64, y: f64) f64 {
    return compareValues(inttypes.wrapS64(x), wrap.fromNumber(y));
}

fn compareU64Double(x: u64, y: f64) f64 {
    return compareValues(inttypes.wrapU64(x), wrap.fromNumber(y));
}

/// Render through the abstract type's own `tostring`, which raises.
fn render(at: *const AbstractType, p: *const anyopaque, b: *buffers.Buffer) !void {
    try at.tostring.?(@constCast(p), @ptrCast(b));
}

fn bufferIs(b: *buffers.Buffer, expected: []const u8) bool {
    const count: usize = @intCast(b.count);
    return count == expected.len and std.mem.eql(u8, b.slice()[0..count], expected);
}

fn eval(source: [*:0]const u8) repr.Value {
    var out: repr.Value = undefined;
    expect(core_env.dostring(environment, source, "inttypes-contract", &out) == 0);
    return out;
}

/// The hash folds the two halves together, so it is stable and the same for
/// either of the two types over the same bits.
fn theHash() void {
    var a: i64 = 0;
    var b: i64 = 1;
    var low: i64 = std.math.minInt(i64);
    var u: u64 = 1;

    expect(s64_type.hash.?(&a, @sizeOf(i64)) == 0);
    expect(s64_type.hash.?(&b, @sizeOf(i64)) ==
        u64_type.hash.?(&u, @sizeOf(u64)));
    expect(s64_type.hash.?(&low, @sizeOf(i64)) == std.math.minInt(i32));
    expect(s64_type.hash.?(&b, @sizeOf(i64)) !=
        s64_type.hash.?(&a, @sizeOf(i64)));

    // Values differing only in the high word still separate, which is what the
    // fold is for.
    var high: i64 = @as(i64, 1) << 32;
    expect(s64_type.hash.?(&high, @sizeOf(i64)) == 1);
}

fn theAbstractCompare() void {
    var s_small: i64 = -5;
    var s_big: i64 = 5;
    var s_min: i64 = std.math.minInt(i64);
    var s_max: i64 = std.math.maxInt(i64);
    var u_small: u64 = 5;
    var u_big: u64 = std.math.maxInt(u64);

    expect(s64_type.compare.?(&s_small, &s_big) == -1);
    expect(s64_type.compare.?(&s_big, &s_small) == 1);
    expect(s64_type.compare.?(&s_big, &s_big) == 0);
    expect(s64_type.compare.?(&s_min, &s_max) == -1);
    expect(s64_type.compare.?(&s_max, &s_min) == 1);

    // The unsigned comparison must not borrow the signed ordering.
    expect(u64_type.compare.?(&u_small, &u_big) == -1);
    expect(u64_type.compare.?(&u_big, &u_small) == 1);
    expect(u64_type.compare.?(&u_big, &u_big) == 0);

    var high_bit: u64 = @as(u64, 1) << 63;
    var one: u64 = 1;
    expect(u64_type.compare.?(&high_bit, &one) == 1);
}

fn theSignedAgainstDoubles() void {
    const inf = std.math.inf(f64);
    const nan = std.math.nan(f64);

    // Inside the double's contiguous integer range the comparison is exact.
    expect(compareS64Double(0, 0.0) == 0);
    expect(compareS64Double(5, 5.0) == 0);
    expect(compareS64Double(5, 5.5) == -1);
    expect(compareS64Double(6, 5.5) == 1);
    expect(compareS64Double(-5, -5.0) == 0);
    expect(compareS64Double(-6, -5.5) == -1);
    expect(compareS64Double(-5, -5.5) == 1);

    // NaN compares equal to everything; see the header comment.
    expect(compareS64Double(0, nan) == 0);
    expect(compareS64Double(std.math.maxInt(i64), nan) == 0);

    // Infinities sit outside every integer.
    expect(compareS64Double(std.math.maxInt(i64), inf) == -1);
    expect(compareS64Double(std.math.minInt(i64), -inf) == 1);

    // Beyond 2^53 the integer cannot be widened without rounding, so the
    // double is narrowed instead.
    expect(compareS64Double(std.math.maxInt(i64), 1e300) == -1);
    expect(compareS64Double(std.math.minInt(i64), -1e300) == 1);
    expect(compareS64Double(std.math.maxInt(i64), 9.3e18) == -1);
    expect(compareS64Double(std.math.minInt(i64), -9.3e18) == 1);

    // 2^53 itself is the edge of the exact range.
    expect(compareS64Double(9007199254740992, 9007199254740992.0) == 0);
    expect(compareS64Double(9007199254740993, 9007199254740992.0) == 1);
    expect(compareS64Double(-9007199254740993, -9007199254740992.0) == -1);

    // 2^63 is the first double past the range, and it orders above every box
    // from either side. The doubles either side of it are the bracket.
    const two_63 = 9223372036854775808.0;
    expect(compareS64Double(1, two_63) == -1);
    expect(compareS64Double(std.math.maxInt(i64), two_63) == -1);
    expect(compareValues(wrap.fromNumber(two_63), inttypes.wrapS64(1)) == 1);
    expect(compareS64Double(1, two_63 - 1024) == -1);
    expect(compareS64Double(1, 2 * two_63) == -1);
    // Below 2^63 the double is still inside the range and compared exactly.
    expect(compareS64Double(std.math.maxInt(i64), 4611686018427387904.0) == 1);
}

fn theUnsignedAgainstDoubles() void {
    const inf = std.math.inf(f64);
    const nan = std.math.nan(f64);
    const max = std.math.maxInt(u64);

    expect(compareU64Double(0, 0.0) == 0);
    expect(compareU64Double(5, 5.0) == 0);
    expect(compareU64Double(5, 5.5) == -1);
    expect(compareU64Double(6, 5.5) == 1);

    // Every unsigned value is above every negative double, including zero
    // against a small negative, which is the case a naive cast gets wrong.
    expect(compareU64Double(0, -0.5) == 1);
    expect(compareU64Double(0, -1e300) == 1);
    expect(compareU64Double(max, -1.0) == 1);

    expect(compareU64Double(0, nan) == 0);
    expect(compareU64Double(max, nan) == 0);
    expect(compareU64Double(max, inf) == -1);
    expect(compareU64Double(max, 1e300) == -1);

    expect(compareU64Double(9007199254740992, 9007199254740992.0) == 0);
    expect(compareU64Double(9007199254740993, 9007199254740992.0) == 1);

    // 2^64 is the first double past the range, from either side, bracketed by
    // the double below it and 2^65.
    const two_64 = 18446744073709551616.0;
    expect(compareU64Double(1, two_64) == -1);
    expect(compareU64Double(max, two_64) == -1);
    expect(compareValues(wrap.fromNumber(two_64), inttypes.wrapU64(1)) == 1);
    expect(compareU64Double(1, two_64 - 2048) == -1);
    expect(compareU64Double(1, 2 * two_64) == -1);
}

/// The two 64-bit types against each other, where neither can be widened into
/// the other without losing a value.
fn theTwoTypesAgainstEachOther() void {
    const s_neg = inttypes.wrapS64(-1);
    const s_zero = inttypes.wrapS64(0);
    const s_max = inttypes.wrapS64(std.math.maxInt(i64));
    const u_zero = inttypes.wrapU64(0);
    const u_small = inttypes.wrapU64(1);
    const u_huge = inttypes.wrapU64(@as(u64, std.math.maxInt(i64)) + 1);
    const u_max = inttypes.wrapU64(std.math.maxInt(u64));

    // A negative signed value is below every unsigned value.
    expect(compareValues(s_neg, u_zero) == -1);
    expect(compareValues(s_neg, u_max) == -1);
    expect(compareValues(u_zero, s_neg) == 1);
    expect(compareValues(u_max, s_neg) == 1);

    // An unsigned value above INT64_MAX is above every signed value.
    expect(compareValues(s_max, u_huge) == -1);
    expect(compareValues(u_huge, s_max) == 1);
    expect(compareValues(s_max, u_max) == -1);

    // Inside the overlap the ordering is ordinary.
    expect(compareValues(s_zero, u_zero) == 0);
    expect(compareValues(s_zero, u_small) == -1);
    expect(compareValues(u_small, s_zero) == 1);
    expect(compareValues(s_max, inttypes.wrapU64(std.math.maxInt(i64))) == 0);
}

fn theFormatters() !void {
    const b: *buffers.Buffer = buffers.new(0);

    var s: i64 = 0;
    try render(s64_type, &s, b);
    expect(bufferIs(b, "0"));

    b.count = 0;
    s = std.math.minInt(i64);
    try render(s64_type, &s, b);
    expect(bufferIs(b, "-9223372036854775808"));

    b.count = 0;
    s = std.math.maxInt(i64);
    try render(s64_type, &s, b);
    expect(bufferIs(b, "9223372036854775807"));

    // The unsigned formatter must not print the high bit as a sign.
    b.count = 0;
    var u: u64 = std.math.maxInt(u64);
    try render(u64_type, &u, b);
    expect(bufferIs(b, "18446744073709551615"));

    // Formatting appends rather than replacing.
    b.count = 0;
    _ = buffers.pushCstringAbi(b, "n=");
    u = 42;
    try render(u64_type, &u, b);
    expect(bufferIs(b, "n=42"));
}

/// Floored division and modulo, through the `div` and `mod` methods.
///
/// `INT64_MIN / -1` has no representable result, and all six methods refuse
/// it with one message. The refusal is asserted at the end, because it is the
/// one input the division has no result to render for.
fn theFlooredDivision() !void {
    const cases = [_]struct { [:0]const u8, []const u8 }{
        // Floored division rounds toward negative infinity, unlike `/`.
        .{ "(div (int/s64 7) (int/s64 2))", "3" },
        .{ "(div (int/s64 -7) (int/s64 2))", "-4" },
        .{ "(div (int/s64 7) (int/s64 -2))", "-4" },
        .{ "(div (int/s64 -7) (int/s64 -2))", "3" },
        .{ "(div (int/s64 8) (int/s64 2))", "4" },
        .{ "(div (int/s64 -8) (int/s64 2))", "-4" },
        .{ "(div (int/s64 0) (int/s64 -3))", "0" },
        // `rdiv` swaps the operands.
        .{ "(:rdiv (int/s64 2) (int/s64 -7))", "-4" },
        // Floored modulo takes the sign of the divisor, unlike `%`.
        .{ "(mod (int/s64 7) (int/s64 2))", "1" },
        .{ "(mod (int/s64 -7) (int/s64 2))", "1" },
        .{ "(mod (int/s64 7) (int/s64 -2))", "-1" },
        .{ "(mod (int/s64 -7) (int/s64 -2))", "-1" },
        .{ "(mod (int/s64 -8) (int/s64 2))", "0" },
        .{ "(% (int/s64 -7) (int/s64 2))", "-1" },
        // A zero divisor returns the dividend rather than raising.
        .{ "(mod (int/s64 7) (int/s64 0))", "7" },
        .{ "(mod (int/s64 -7) (int/s64 0))", "-7" },
        .{ "(:rmod (int/s64 2) (int/s64 -7))", "1" },
        // The extremes still divide when the result is representable.
        .{ "(div (int/s64 \"-9223372036854775808\") (int/s64 2))", "-4611686018427387904" },
        .{ "(mod (int/s64 \"-9223372036854775808\") (int/s64 3))", "1" },
    };

    for (cases) |case| {
        var result: repr.Value = undefined;
        expect(core_env.dostring(environment, case[0].ptr, "inttypes-contract", &result) == 0);
        expect(inttypes.isInt(result) == constants.IntType.s64);

        const b: *buffers.Buffer = buffers.new(0);
        try render(s64_type, wrap.toAbstract(result), b);
        expect(bufferIs(b, case[1]));
    }

    // Dividing by zero raises, while the modulo above does not.
    //
    // Through `vm_entry.pcall` rather than `harness.raised`, because `div` is
    // not an nfunction. It is a Janet function that dispatches to the abstract
    // type's `div` method, so there is no `raise.NFunction` to call and no
    // error to catch: the raise happens inside the interpreter, and a
    // protected call is the instrument for that. `harness.core("div")` fails
    // its type assertion.
    const closure = eval("(fn [] (div (int/s64 1) (int/s64 0)))");
    expect(vm_entry.pcall(
        wrap.toFunction(closure),
        &.{},
        null,
    ).signal == abi.Signal.@"error");

    // `INT64_MIN / -1`, through every method that can reach it. Each is a
    // Janet-level dispatch for the reason above, so each is a protected call.
    for ([_][:0]const u8{
        "(fn [] (/ (int/s64 \"-9223372036854775808\") (int/s64 -1)))",
        "(fn [] (% (int/s64 \"-9223372036854775808\") (int/s64 -1)))",
        "(fn [] (div (int/s64 \"-9223372036854775808\") (int/s64 -1)))",
        "(fn [] (mod (int/s64 \"-9223372036854775808\") (int/s64 -1)))",
        "(fn [] (:rdiv (int/s64 -1) (int/s64 \"-9223372036854775808\")))",
        "(fn [] (:rmod (int/s64 -1) (int/s64 \"-9223372036854775808\")))",
    }) |source| {
        const fn_value = eval(source.ptr);
        expect(vm_entry.pcall(wrap.toFunction(fn_value), &.{}, null).signal == abi.Signal.@"error");
    }

    // And the neighbouring divisors, which are representable and must not be
    // caught by the guard.
    for ([_][:0]const u8{
        "(fn [] (div (int/s64 \"-9223372036854775808\") (int/s64 1)))",
        "(fn [] (div (int/s64 \"-9223372036854775807\") (int/s64 -1)))",
    }) |source| {
        const fn_value = eval(source.ptr);
        expect(vm_entry.pcall(wrap.toFunction(fn_value), &.{}, null).signal == abi.Signal.ok);
    }
}

/// An operand the boxed types cannot convert refuses *catchably*.
///
/// Asserted from Janet rather than through `harness.raised` for the reason
/// above: this is reached by the interpreter's binop fallback rather than
/// called directly. What is under test is not that the conversion refuses,
/// which `theSignedAgainstDoubles` covers, but that the refusal *arrives*.
///
/// A `Box(T).unwrap` bound to a reporting form of the conversion from inside a
/// raise-capable method would end the process with `a raise was reported
/// across the C ABI and never consumed` instead of raising, and it would be
/// hidden behind a comptime alias. `res/check/swallowed.janet` is what looks
/// for that shape.
fn anUnconvertibleOperandRefusesCatchably() void {
    const cases = [_][*:0]const u8{
        "(fn [] (+ (int/s64 1) {}))",
        "(fn [] (+ (int/s64 1) \"x\"))",
        "(fn [] (* (int/u64 2) [1]))",
        "(fn [] (- (int/s64 1) !{}))",
        // The `r`-prefixed methods, which the fallback reaches when the boxed
        // integer is the right operand.
        "(fn [] (+ {} (int/s64 1)))",
        // The in-place folds, which read through the same alias.
        "(fn [] (band (int/u64 1) :nope))",
    };
    for (cases) |source| {
        const closure = eval(source);
        const resumed = vm_entry.pcall(
            wrap.toFunction(closure),
            &.{},
            null,
        );
        expect(resumed.signal == abi.Signal.@"error");
        // And the payload is the conversion's own message, which is what says
        // the refusal travelled rather than being manufactured downstream.
        expect(harness.isType(resumed.value, repr.Tag.string));
        const message = wrap.toString(resumed.value);
        const length: usize = strings.head(message).length;
        expect(std.mem.indexOf(u8, message[0..length], "can not convert") != null);
    }
}

// ==========================================================================
// Entry
// ==========================================================================

fn body() !void {
    theHash();
    theAbstractCompare();
    theSignedAgainstDoubles();
    theUnsignedAgainstDoubles();
    theTwoTypesAgainstEachOther();
    try theFormatters();
    try theFlooredDivision();
    anUnconvertibleOperandRefusesCatchably();
}

pub fn run() void {
    harness.init();
    environment = harness.coreEnv();

    const fun = eval("(fn [a b] (compare a b))");
    compare_fn = wrap.toFunction(fun);
    gc_alloc.gcroot(fun);

    body() catch @panic("inttypes: a kernel raised unexpectedly");

    vm_lifecycle.deinit();
}
